key-rotator/manifest/manifest.go

```diff
@@ -30,6 +37,68 @@ type DataShareProcessorSpecificManifest struct {
 	PacketEncryptionKeyCSRs PacketEncryptionKeyCSRs `json:"packet-encryption-keys"`
 }
 
+// UpdateKeysConfig configures an UpdateKeys operation.
+type UpdateKeysConfig struct {
+	BatchSigningKey         key.Key // the key used for batch signing operations
+	BatchSigningKeyIDPrefix string  // the key ID prefix to use for batch signing keys
+
+	PacketEncryptionKey         key.Key // the key used for packet encryption operations
+	PacketEncryptionKeyIDPrefix string  // the key ID prefix to use for packet encryption keys
+	PacketEncryptionKeyCSRFQDN  string  // the FQDN to specify for packet encryption key CSRs
+}
+
+func (m DataShareProcessorSpecificManifest) UpdateKeys(cfg UpdateKeysConfig) (DataShareProcessorSpecificManifest, error) {
+	// Copy the current manifest, clearing any existing batch signing/packet encryption keys.
+	newM := m
+	newM.BatchSigningPublicKeys, newM.PacketEncryptionKeyCSRs = BatchSigningPublicKeys{}, PacketEncryptionKeyCSRs{}
+
+	// Update batch signing key.
+	for _, v := range cfg.BatchSigningKey {
+		kid := cfg.BatchSigningKeyIDPrefix
+		if !v.CreationTime.IsZero() {
```

Comment:
In what case would a `key.Key` version not have a `CreationTime`? My guess is it's in the case where we are migrating from a non-versioned key discovered in Kubernetes secrets? If I'm right, then there should only ever be one key version in a key with a zero `CreationTime`. This function already seems to make that assumption, and if I'm reading it right, the first zero `CreationTime` encountered when iterating the slice will "win" and be present in the manifest; others are silently dropped. We should have an explicit check for multiple versions with no `CreationTime` and return an error in that case.

Reply:
You're correct about when `key.Key` would have a zero `CreationTime` -- if and only if it is a "migrated" key version from the currently-deployed serialization strategy.

There are a number of sharp edges/risky cases that "shouldn't" be possible, but I strongly agree that it would be good to validate. OTOH, I am still up in the air about the exact validations & where they fit into the larger program logic.

At the least, I want to implement the following validations:
* Keys must either have no versions, or exactly one primary version.
* Post-rotation, keys must have at least one version.
* When being used to update keys in a manifest, keys must have at least one version.
* When being used to update keys in a manifest, keys must have versions with distinct creation times.
* When being used to update keys in a manifest, existing key versions' public portions should match with the public keys written to the manifest, matching by creation time/key ID (both of which would provide the same matching).
* Post-update, manifests should have exactly one packet encryption key version.
* Post-update, manifests should have at least one batch signing key version.

Any more that you can think of?

If you are OK with it, I would like to wait to add this kind of validation until I have the first version of the "main" functionality written tying everything together -- I want to think through where these validations best fit in a whole-program sense, which should be tied directly to data types vs included in program logic, which should be bypassable by passing the right --unsafe_no_really_force_rotation flag(s) vs should really always be applied, etc.

Reply:
I'm primarily worried with avoiding any silent "failures", scenarios where malformed or unexpected keys cause us to use whatever value an iterator coughs up first without errors being raised or anomalies being logged. I'm OK with deferring that to later since a good place to centralize that sort of validation might come along as more pieces of the overall system fall into place.

Reply:
I added the list above to https://github.com/abetterinternet/prio-server/issues/24 to help ensure it doesn't get lost.